Constraint template filters

lsviben opened the pull request:
**What this PR does / why we need it**:
Implements constraint template filters which are used in Kubermatic EE to allow users to filter out CTs for different clusters.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6540 

**Does this PR introduce a user-facing change?**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If  no release note is required, just write "NONE".
-->
```release-note
Kubermatic Constraint Templates now support cluster filtering to allow users to manage the CT's and policies across different clusters more easily. The feature is available for Kubermatic EE. 
```

lsviben committed:
CT cluster filtering

lsviben committed:
ct filters test

lsviben committed:
gen swagger and api clients

lsviben commented:
/retest

lsviben committed:
lint fix

moadqassem commented on pkg/controller/seed-controller-manager/constraint-template-controller/wrappers_ce.go:
```diff
@@ -0,0 +1,36 @@
+// +build !ee
+
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
```
```suggestion
Copyright 2021 The Kubermatic Kubernetes Platform contributors.
```

moadqassem commented on pkg/controller/seed-controller-manager/constraint-template-controller/wrappers_ce.go:
```diff
@@ -0,0 +1,36 @@
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *reconciler) getClustersForConstraintTemplate(ctx context.Context, ct *kubermaticv1.ConstraintTemplate) (*kubermaticv1.ClusterList, error) {
```
This function is only used once and it doesn't do anything special other than listing the clusters, why not just use the listing inline instead of having a dedicated method for it :-)

moadqassem commented on pkg/controller/seed-controller-manager/constraint-template-controller/wrappers_ee.go:
```diff
@@ -0,0 +1,30 @@
+// +build ee
+
+/*
+Copyright 2020 The Kubermatic Kubernetes Platform contributors.
```
```suggestion
Copyright 2021 The Kubermatic Kubernetes Platform contributors.
```

moadqassem commented on pkg/controller/seed-controller-manager/constraint-template-controller/wrappers_ee.go:
```diff
@@ -0,0 +1,30 @@
+	eeseedctrlmgr "k8c.io/kubermatic/v2/pkg/ee/cmd/seed-controller-manager"
+)
+
+func (r *reconciler) getClustersForConstraintTemplate(ctx context.Context, ct *kubermaticv1.ConstraintTemplate) (*kubermaticv1.ClusterList, error) {
```
Similar to what I have said earlier, why wrapping is needed?

moadqassem commented on pkg/crd/kubermatic/v1/cluster.go:
```diff
@@ -44,6 +44,20 @@ const (
 	// CredentialPrefix is the prefix used for the secrets containing cloud provider crednentials.
 	CredentialPrefix = "credential"
+
+	AlibabaProvider      = "alibaba"
```
We already have an enum for all the cloud providers that we support:
https://github.com/kubermatic/kubermatic/blob/master/pkg/provider/types.go#L49-L62

moadqassem commented on pkg/crd/kubermatic/v1/cluster.go:
```diff
@@ -730,3 +744,39 @@ func (cluster *Cluster) IsOpenshift() bool {
 	return !cluster.IsOpenshift()
 }
+
+func (cluster *Cluster) GetProviderName() string {
```
You don't need this method as we have different utility functions to find the cluster cloud provider. You can use this function here  https://github.com/kubermatic/kubermatic/blob/master/pkg/provider/types.go#L450 to get the cloud provider out of a subset, or directly this function here based in the cloud spec: https://github.com/kubermatic/kubermatic/blob/master/pkg/provider/types.go#L398

moadqassem commented on pkg/ee/cmd/seed-controller-manager/wrappers.go:
```diff
@@ -51,3 +54,10 @@ func SeedGetterFactory(ctx context.Context, client ctrlruntimeclient.Client, see
 }
+
+// GetClustersForConstraintTemplate gets clusters for the CT by using the CT selector to filter out unselected clusters
+func GetClustersForConstraintTemplate(ctx context.Context, client ctrlruntimeclient.Client,
```
Again, why wrapping?

moadqassem commented on pkg/ee/constraint-template-controller/controller.go:
```diff
@@ -0,0 +1,76 @@
+/*
+                  Kubermatic Enterprise Read-Only License
+                         Version 1.0 ("KERO-1.0”)
+                     Copyright © 2020 Loodse GmbH
```
```suggestion
                     Copyright © 2021 Kubermatic GmbH
```

moadqassem commented on pkg/ee/constraint-template-controller/controller.go:
```diff
@@ -0,0 +1,76 @@
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/sets"
```
Please fix the regrouping

moadqassem commented on pkg/ee/constraint-template-controller/controller.go:
```diff
@@ -0,0 +1,76 @@
+// +build ee
```
Why this file is called controller? I don't see any reconciling loop logic in here. The functions listed below are kinda representing utility functions

moadqassem commented on pkg/ee/constraint-template-controller/controller.go:
```diff
@@ -0,0 +1,76 @@
+
+	ctLabelSelector, err := v1.LabelSelectorAsSelector(&ct.Spec.Selector.LabelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("Error converting CT label selector to selector: %v", err)
```
Can you please make the error message clearer as this doesn't quite have a meaningful error.

moadqassem commented on pkg/ee/constraint-template-controller/controller_test.go:
```diff
@@ -0,0 +1,154 @@
+/*
+                  Kubermatic Enterprise Read-Only License
+                         Version 1.0 ("KERO-1.0”)
+                     Copyright © 2020 Loodse GmbH
```
```suggestion
                     Copyright © 2021 Kubermatic GmbH
```

lsviben committed:
code review fixes

lsviben committed:
rename file

lsviben committed:
using existing provider name getters

lsviben committed:
fix 2020 -> 2021

moadqassem reviewed (commented):
@lsviben Thank you very much for addressing the comments. One last thing open is the wrapping stuff, can you elaborate why this is needed? 

/retest

lsviben commented:
> @lsviben Thank you very much for addressing the comments. One last thing open is the wrapping stuff, can you elaborate why this is needed?
> 
> /retest

Thanks for the review.

I wrapped the `getClustersForConstraintTemplate(..)`  so we can have different builds for `ee` and `ce` versions, so that the filtering is done only for the `ee` version, while the `ce` version just does a regular list. Not sure how to do it differently

lsviben committed:
boilerplate Kubmermatic to Loodse, fix in other pr

moadqassem reviewed (commented):
/approve
/lgtm

lsviben commented:
/retest

lsviben commented:
/retest
